### PR TITLE
Reimplement select multi widgets using recycler view

### LINKED ISSUE #2494: Long lists of choices take a long time to render
#### Problem description
If the list of choices in a select one or select multiple question is long, it takes a long time to display that question because all `TextView`s are computed and held in memory. 

#### Steps to reproduce the problem
Load a form with a select question that has a long list of choices. These can come from a static choice list, a dynamic choice list or an external DB (using "search/pulldata" or "external itemsets").

#### Expected behavior
Views are only built for the currently-displayed items and a few before and after.

#### Other information 
As @shobhitagarwal1612 described in https://github.com/opendatakit/collect/issues/1254, this is probably a compelling place to use a `RecyclerView`.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #2494 

#### What has been done to verify that this works as intended?
I tested the attached forms.

#### Why is this the best possible solution? Were any other approaches considered?
it's a continuation of https://github.com/opendatakit/collect/pull/2513 using the apporach which improves the performance a lot.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
The risk here is related to all selectOne and selectMultiple widgets.

As I said it's continuation of https://github.com/opendatakit/collect/pull/2513 which has been already tested and merged. Unfortunatelly I had to refactor code in both select one and multiple in order to finish this implementation. That means we need to test the same scenarios like it was in https://github.com/opendatakit/collect/pull/2513 and additionally the same but for select multiple.

#### Do we need any specific form for testing your changes? If so, please attach one.
[select_one_external.zip](https://github.com/opendatakit/collect/files/2396432/select_one_external.zip)

[selectMulit7ThousandItems.zip]
(https://github.com/opendatakit/collect/files/2396418/selectMulit7ThousandItems.zip)

https://drive.google.com/file/d/1jZg3ye-27zkZeBhxUIzDNyoAusjt2UeF/view?usp=sharing

[AOOP.xml.txt](https://github.com/opendatakit/collect/files/2396408/AOOP.xml.txt) - because in my last commit I fixed a bug which is visible on master barcnch caused by #2513 

and AllWidgets form.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by grzesiek2010** [comment]
@mmarciniak90 @lognaturel 
In this pr, I fixed the second issue mentioned https://github.com/opendatakit/collect/pull/2513#issuecomment-421320889 I mean not visible scrollbar.

When it comes to the first one it's not a big problem. I think the cause might be related to https://github.com/opendatakit/collect/issues/2537 since it's visible only after rotating the screen or returning from HierarchyView (when double instantiation of widgets takes place). So let's wait maybe https://github.com/opendatakit/collect/pull/2538 will fix it if not I'll take a deeper look.

**Comment by mmarciniak90** [comment]
**Tested with success!**

Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

Verified forms:
* All widgets
* AOOP
* items
* select_one_external
* selectMulti 

Problem with the scrollbar in form `select_one_external` is fixed.

The problem with visibility of widgets after changing device orientation will be fixed as a part of #2537 
- select_one type with no appearance, 4 image choices
- select_one type with compact-2 appearance, 4 image choices
- select_one type with quickcompact-2 appearance, 4 image choices
- select_one with type image-map appearance
- select_multiple type with compact-2 appearance, 4 image choices
- select_multiple with type image-map appearance

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"

**Comment by grzesiek2010** [comment]
:+1: 
@lognaturel please take a look and if it's ok we can merge and announce the code freeze :smile: 

as I said and @mmarciniak90 pointed out one small bug is still visible but let's put it on hold (see my comment above).

You may be a little bit disappointed because I didn't fix dependence between the widget and the ViewHolder https://github.com/opendatakit/collect/pull/2513#issuecomment-420105432
maybe there is a better way but it would need a bigger refactoring. For now, I need passing a widget to Viewholder in order to perform operations like
https://github.com/opendatakit/collect/pull/2547/files#diff-16abb480481e8a0a0d07cf1a8bee3aa2R120
https://github.com/opendatakit/collect/pull/2547/files#diff-d38f3b3845ce85a71ff2999805e14818R50. 
It's just code style but performance is much much better so I hope you are satisfied.

**Comment by lognaturel** [comment]
I’ve only taken a quick look so far and would like to review more deeply after the TSC call before merging but I see no big issues that need addressing now!

File: collect_app/src/main/java/org/odk/collect/android/adapters/AbstractSelectListAdapter.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    public int getItemCount() {
+        return filteredItems.size();
+    }
+
+    @Override
+    public Filter getFilter() {
+        return new Filter() {
+            @Override
+            protected FilterResults performFiltering(CharSequence charSequence) {
+                String searchStr = charSequence.toString().toLowerCase(Locale.US);
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/adapters/AbstractSelectListAdapter.java:71]: What are the implications of using `Locale.US` on a non-English language, particularly on a script other than the Latin alphabet?

File: collect_app/src/main/java/org/odk/collect/android/widgets/SelectWidget.java [hunk 2]
```diff
@@ -165,4 +161,22 @@ public void addMediaFromChoice(MediaLayout mediaLayout, int index, TextView text
...
+    protected RecyclerView setUpRecyclerView() {
+        RecyclerView recyclerView = (RecyclerView) LayoutInflater.from(getContext()).inflate(R.layout.recycler_view, null); // keep in an xml file to enable the vertical scrollbar
+        recyclerView.addItemDecoration(new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL));
+
+        return recyclerView;
+    }
+
+    void adjustRecyclerViewSize(AbstractSelectListAdapter adapter, RecyclerView recyclerView) {
+        if (adapter.getItemCount() > MAX_ITEMS_WITHOUT_SCREEN_BOUND) {
+            // Only let the RecyclerView take up 80% of the screen height in order to speed up loading if there are many items
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/SelectWidget.java]: This should be 90% to match the code.

**Review by lognaturel — APPROVED** [review]

**Comment by lognaturel** [comment]
> You may be a little bit disappointed because I didn't fix dependence between the widget and the ViewHolder

A little bit! But I get it. Maybe slight improvement would be to use a getter to get the `items` when needed rather than also having a field for those since they're members of the widget. But it's not urgent.

The only change I think is necessary before merge is to fix the comment. @grzesiek2010 I'll wait a couple of hours to see if you're around to address that. If not, I'll do it and then merge.

I also have a question about the locale used to alter casing but I don't think this is urgent. I think that if it's done consistently on both sides it's likely ok. I'm guessing the risk is false positives more than false negatives (that is, finding things that aren't actually matches rather than missing matches).

File: collect_app/src/main/java/org/odk/collect/android/widgets/SelectWidget.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/widgets/SelectWidget.java]: Good catch :smile: Done.

File: collect_app/src/main/java/org/odk/collect/android/adapters/AbstractSelectListAdapter.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/adapters/AbstractSelectListAdapter.java:71]: Yes, in this case, it's rather useless but lint requires it in all such cases.

**Pull request merged**